File: backend/src/service/milestone.js
const Milestone = require('../model').Milestone;
const Issue = require('../model').Issue;

module.exports = {
  create: async ({ title, content, deadline } = {}) => {
    if (!title || !content || !deadline) {
      return { error: '정보가 부족합니다' };
    }
    const milestone = await Milestone.create({
      title,
      content,
      deadline,
      is_opened: true,
    });
    return milestone;
  },

  read: async () => {
    const milestones = await Milestone.findAll({
      include: [Issue],
    });

    const milestone = milestones.map((mile) => {
      const open = mile.dataValues.Issues.filter((issue) => issue.is_opened)
        .length;
      const total = mile.dataValues.Issues.length;
      mile.dataValues['openCount'] = open;
      mile.dataValues['totalCount'] = total;
      delete mile.dataValues.Issues;
      return mile.dataValues;
    });

    return { milestones: milestone };
  },

  update: async ({ id, title, deadline, content } = {}) => {
    if (!id) {
      return { error: '정보가 부족합니다' };
    }

    const [result] = await Milestone.update(
      { title, deadline, content },
      { where: { id } }
    );
    if (result === 1) return { response: true };
    return { response: false };
  },
  remove: async ({ id } = {}) => {
    if (!id) {
      return { error: '정보가 부족합니다' };
    }

    const result = await Milestone.destroy({
      where: { id },
    });
    if (result === 1) return { response: true };
    return { response: false };
  },

  updateState: async ({ id, isOpened } = {}) => {
    if (!id) {
      return { error: '정보가 부족합니다' };
    }
    const milestone = await Milestone.findOne({ where: { id } });

    if (!milestone) {
      return { error: '마일스톤이 존재하지 않습니다' };
    }
    const [result] = await Milestone.update(
      { is_opened: isOpened },
      { where: { id } }
    );
    if (result) {
      return { response: true };
    }
    return { error: 'Milestone 상태 변경 실패' };
  },
};

File: iOS/IssueTracker/IssueTracker/Models/Milestone.swift
//
//  Milestone.swift
//  IssueTracker
//
//  Created by 현기엽 on 2020/11/02.
//

import Foundation

struct Milestones: Codable {
    let milestones: [Milestone]
}

struct Milestone: Model, Codable {
    let id: Int
    var title, content, deadline: String
    let isOpened: Bool
    let openCount: Int?
    let totalCount: Int?
    
    enum CodingKeys: String, CodingKey {
        case id, title, content, deadline
        case isOpened = "is_opened"
        case openCount, totalCount
    }
}

extension Milestone: Hashable {
    static func == (rhs: Milestone, lhs: Milestone) -> Bool {
        return rhs.id == lhs.id &&
            rhs.title == lhs.title &&
            rhs.content == lhs.content &&
            rhs.deadline == lhs.deadline &&
            rhs.isOpened == lhs.isOpened &&
            rhs.openCount == lhs.openCount &&
            rhs.totalCount == lhs.totalCount
    }
}
